### PR TITLE
Add Chapter 1 of the Programming Guide

### PR DESCRIPTION
**Pull request by mbasmanova**

### CODE REVIEW

File: velox/docs/programming-guide/chapter01.rst [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+    nulls->setSize(20);
+
+    > 20, 32
+
+Notice that setSize method takes number of bytes, not number of values of type
+T. Also, Buffer itself is not a template and therefore is not aware of the type
+of values that are being stored in it.
+
+Setting the size to a value that exceeds the Buffer’s capacity results in an
```
Comment (xiaoxmeng) [inline velox/docs/programming-guide/chapter01.rst:79]:
```
nit: s/results in an error/results in a Velox runtime exception error or results in an exception error?/
```

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova Looks great. Thanks!

File: velox/docs/programming-guide/chapter01.rst [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+    > [FLAT BIGINT: 100 elements, 50 nulls]
+
+To make a vector, we use FlatVector<T> class template with T being int64_t
+(64-bit integer). We pass a pointer to MemoryPool, a Type object that describes
+the type of values to store, nulls buffer, number of values, values buffer and
+an empty list of string buffers. Let’s ignore the string buffers for now. We’ll
+discuss these later.
+
+FlatValue<T> class can be used to store values of primitive types. The following
```
Comment (leoluan2009) [inline velox/docs/programming-guide/chapter01.rst:282]: FlatVector<T> ?

File: velox/docs/programming-guide/chapter01.rst [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    auto pool = memory::getDefaultMemoryPool();
+
+`pool` is a std::shared_ptr<velox::memory::MemoryPool>. We can use it to
+allocate buffers.
+
+.. code-block:: c++
+
+    #include "velox/buffer/Buffer.h"
+
+    auto values = AlignedBuffer::allocate<int64_t>(100, pool.get());
```
Comment (aditi-pandit) [inline velox/docs/programming-guide/chapter01.rst:27]: A common question I have received here is how to get this type from the Velox type that we want to create the vector for. Might be worth clarifying this here.

File: velox/docs/programming-guide/chapter01.rst [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+Type                    C++ Type                       Description
+======================  ===========================    ==================
+BOOLEAN                 bool                           A boolean flag: true or false.
+TINYINT                 int8_t                         8-bit integer.
+SMALLINT                int16_t                        16-bit integer.
+INTEGER                 int32_t	                       32-bit integer.
+BIGINT                  int64_t                        64-bit integer.
+REAL                    float                          32-bit floating point number.
+DOUBLE                  double                         64-bit floating point number.
+VARCHAR                 struct StringView              Variable width string.
```
Comment (aditi-pandit) [inline velox/docs/programming-guide/chapter01.rst:295]: Would be worth adding some content on how to handle DATE, TIMESTAMP and DECIMALs to this list.

File: velox/docs/programming-guide/chapter01.rst [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    LOG(INFO) << allNullVector->toString();
+
+    > [FLAT BIGINT: 100 elements, 100 nulls]
+
+However, it is invalid to create a vector with both nulls and values buffers being null.
+
+.. code-block:: c++
+
+    std::make_shared<FlatVector<int64_t>>(
+       pool.get(), BIGINT(), nullptr, 100, nullptr, std::vector<BufferPtr>{});
```
Comment (aditi-pandit) [inline velox/docs/programming-guide/chapter01.rst:326]: Would be great to point out the relation between the C++ type and Velox type here.

**Pull request closed without merging**